### ISSUE TITLE
Should bare tests result in two levels of groups?

### ISSUE THREAD

**Issue by DanTup** [dart-lang/test#836]
When running some flutter tests that in a group like this:

```dart
void main() {
  testWidgets('Hello world test', (WidgetTester tester) async {
    hello_world.main(); // BREAKPOINT1
    await tester.pump();
    expect(find.text('Hello, world!'), findsOneWidget);
  });
}
```

I get back two levels of groups in addition to the suite:

```js
	{
		"group": {
			"id": 2,
			"suiteID": 0,
			"name": null,
			"testCount": 1,
			"url": null
		},
		"type": "group",
		"time": 5608
	},
	{
		"group": {
			"id": 3,
			"suiteID": 0,
			"parentID": 2,
			"name": "-",
			"testCount": 1,
			"line": 60,
			"column": 16,
			"url": "package:flutter_test/src/widget_tester.dart"
		},
		"type": "group",
		"time": 5609
	},
	{
		"test": {
			"id": 4,
			"name": "- Hello world test",
			"suiteID": 0,
			"groupIDs": [
				2,
				3
			],
			"metadata": {
				"skip": false,
				"skipReason": null
			},
			"line": 61,
			"column": 18,
			"url": "package:flutter_test/src/widget_tester.dart"
		},
		"type": "testStart",
		"time": 5609
	}
```

I'm trying to put the results into a tree in VS Code so I'm trying to understand how best to handle each node. If I just render as-is, the user will see a `Suite` node, a name-less `Group` node, a `Group` named `-` and then the `Test` node. However I don't know whether stripping nodes out will cause issues because I'm not sure why the structure is as it is.

So, assumptions/questions:

1. The only tests that will appear directly in a Suite (with no group) are the "loading..." tests?
2. Will all tests always be at least three levels deep (Suite, null Group, and then either a user-named group or a group named `-` for bare tests)?

If so, does it seem like a sensible thing for me to:

1. Always skip over the `null` group, connecting it's child groups directly to the suite
2. Pull any tests in a group named `-` into the parent (which will be the suite, after we do `1`)

(I guess I'm also curious why we have the phantom null groups, and if they're to try and ensure a suite never has tests in it directly, then why is there a second group for the bare tests?)

**Comment by natebosch** [comment]
I'd be curious to know if `testWidgets` is having an impact.

One of the groups might have something to do with the way the test packages considers loading like a test. @grouma - do you have any familiarity with where these might be coming from?

**Comment by DanTup** [comment]
Aha, good call. If I call test normally, it doesn't happen:

<img width="928" alt="screen shot 2018-05-24 at 20 07 24" src="https://user-images.githubusercontent.com/1078012/40506057-20396186-5f8e-11e8-99c7-058a8a1aaf0f.png">

So `testWidgets` seems to be adding an additional unnamed group. I wonder how these appear in IntelliJ (it has a tree?) or whether it does something to hide them..

**Comment by DanTup** [comment]
Also seems like the weird hyphens are being added by Flutter too, since they're not prefixed for the normal test call. I wonder whether maybe flutter's test code was tweaked to improve the console output?

**Comment by DanTup** [comment]
Aha, I found in the `flutter_test` source code `group('-'`!

Looks like it was added a long time ago; there was a comment about it [here](https://github.com/flutter/flutter/pull/3622#issuecomment-215616451). I'll raise an issue in the flutter repo for discussion.

I guess some of the other questions are still valid, though without this extra level (like is it guaranteed that all non-loading tests are always going to be in the `null` group, so that level could be totally skipped).
